org.osgi.test.junit5.cm/src/main/java/org/osgi/test/junit5/cm/ConfigurationExtension.java

```diff
@@ -0,0 +1,407 @@
+					.pid()
+					.equals(WithConfiguration.NOT_SET)) {
+
+					if (valueUsed) {
+						throw new IllegalArgumentException(
+							"@InjectConfiguration - only one of the Fields `value`, `withConfig` or `withFactoryConfig` could be used.");
+					}
+					WithConfiguration wc = ic.withConfig();
+					configuration = handleWithConfiguration(wc, getConfigAdmin(), blockingConfigHandler);
+					withConfigUsed = true;
+				} else if (!ic.withFactoryConfig()
+					.factoryPid()
+					.equals(WithConfiguration.NOT_SET)) {
+
+					if (valueUsed || withConfigUsed) {
+						throw new IllegalArgumentException(
+							"@InjectConfiguration - only one of the Fields `value`, `withConfig` or `withFactoryConfig` could be used.");
+					}
+
+					WithFactoryConfiguration wc = ic.withFactoryConfig();
+					configuration = handleWithFactoryConfiguration(wc);
+					withFactoryConfigUsed = true;
+				}
+				if (!valueUsed && !withConfigUsed && !withFactoryConfigUsed) {
+					throw new IllegalArgumentException(
+						"@InjectConfiguration - one of the Fields `value`, `withConfig` or `withFactoryConfig` must be used.");
+				}
+				return transformConfigToTargetType(configuration, targetType, getConfigAdmin());
+			}
+			return null;
+		} catch (Exception e) {
+
+			throw new ParameterResolutionException("Could not get Configuration from Configuration-Admin", e);
+		}
+	}
+
+	private Object transformConfigToTargetType(Configuration configuration, TargetType targetType,
+		ConfigurationAdmin configurationAdmin) throws Exception {
+
+		if (targetType.matches(Configuration.class)) {
+			return configuration;
+		} else if (targetType.matches(Optional.class, Configuration.class)) {
+			return Optional.ofNullable(configuration);
+		} else if (targetType.matches(Map.class, String.class, Object.class)) {
+			return Dictionaries.asMap(configuration.getProperties());
+		} else if (targetType.matches(Dictionary.class, String.class, Object.class)) {
+			return configuration.getProperties();
+		}
+		throw new ParameterResolutionException("Bad Parameter-Type");
+	}
+
+	@Override
+	public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext)
+		throws ParameterResolutionException {
+
+		if (parameterContext.isAnnotated(InjectConfiguration.class)) {
+			Parameter parameter = parameterContext.getParameter();
+			if (parameter.getType()
+				.isAssignableFrom(Configuration.class)
+				|| parameter.getType()
+					.isAssignableFrom(Optional.class)
+					&& (((ParameterizedType) parameter.getParameterizedType()).getActualTypeArguments()[0]
+						.equals(Configuration.class))
+				|| parameter.getType()
+					.isAssignableFrom(Map.class)
+				|| parameter.getType()
+					.isAssignableFrom(Dictionary.class)) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	@Override
+	public void beforeAll(ExtensionContext extensionContext) throws Exception {
+		bundleContext = FrameworkUtil.getBundle(extensionContext.getRequiredTestClass())
```

Comment:
You should not get the BundleContext directly. Use BundleContextExtension to get a BundleContext if needed.

Reply:
done